src/server/gc_worker/compaction_runner.rs

```diff
@@ -0,0 +1,728 @@
+                        if candidate.score > top.0.score {
+                            candidates_heap.pop(); // Remove lowest score
+                            candidates_heap.push(Reverse(candidate)); // Add new candidate
+                        }
+                    }
+                }
+                Ok(None) => {} // No compaction needed
+                Err(e) => {
+                    warn!(
+                        "failed to evaluate region {} as compaction candidate: {:?}",
+                        region.get_id(),
+                        e
+                    );
+                }
+            }
+
+            match next_key {
+                Some(key) => current_key = key,
+                None => break,
+            }
+        }
+
+        // Convert heap to sorted vector (highest score first)
+        let mut candidates: Vec<CompactionCandidate> = candidates_heap
+            .into_sorted_vec()
+            .into_iter()
+            .map(|reverse| reverse.0)
+            .collect();
+        candidates.reverse(); // Min-heap gives us lowest first, we want highest first
+
+        // Record MVCC stats for top 10 candidates and log details
+        for (rank, candidate) in candidates.iter().take(10).enumerate() {
+            AUTO_COMPACTION_CANDIDATE_HISTOGRAM_VEC
+                .total
+                .observe(candidate.num_total_entries as f64);
+            AUTO_COMPACTION_CANDIDATE_HISTOGRAM_VEC
+                .tikv_estimated_discardable
+                .observe(candidate.num_discardable as f64);
+            AUTO_COMPACTION_CANDIDATE_HISTOGRAM_VEC
+                .rocksdb_tombstones
+                .observe(candidate.num_tombstones as f64);
+            AUTO_COMPACTION_CANDIDATE_HISTOGRAM_VEC
+                .tikv_rows
+                .observe(candidate.num_rows as f64);
+
+            info!("top compaction candidate";
+                "rank" => rank + 1,
+                "region_id" => candidate.region_id,
+                "score" => candidate.score,
+                "total_entries" => candidate.num_total_entries,
+                "tikv_estimated_discardable" => candidate.num_discardable,
+                "rocksdb_tombstones" => candidate.num_tombstones,
+                "tikv_rows" => candidate.num_rows,
+            );
+        }
+
+        info!("collected {} compaction candidates", candidates.len());
+        fail_point!("gc_worker_auto_compaction_candidates_collected");
+        Ok(candidates)
+    }
+
+    /// Compact candidates and return elapsed time
+    fn compact_candidates(
+        &mut self,
+        candidates: Vec<CompactionCandidate>,
+        config: &GcConfig,
+    ) -> Option<Duration> {
+        let start_time = Instant::now();
+        let mut processed_count = 0;
+        fail_point!("gc_worker_auto_compaction_start_compacting");
+
+        for candidate in candidates {
+            if self.check_stopped() {
```

Comment:
Where is the logic `Compacts candidates as many as possible within time limit, instead of compacting top_n`

Reply:
As commented, now top_n is only used for limiting memory usage, in case single host has too many regions. Now, this heap has 300 capacity by default. I don't think we can compact 300 ranges in 5 minutes. Capacity increases as interval increases, we can change it to interval_as_secs * 10, if you want